File: Assets/Scripts/Framework/Command/CommandContainerBase.cs
using System;

namespace Launch
{
	public abstract class CommandContainerBase : CommandBase
	{
		public event Action<CommandBase> On_ChildDone;

		/// <summary>
		/// 当前子对象执行完成
		/// </summary>
		/// <param name="command">Command.</param>
		public virtual void OnChildDone(CommandBase command)
		{
			On_ChildDoneCallback (command);
			OnChildDestroy (command);
		}

		/// <summary>
		/// 当前子对象执行完成的回调
		/// </summary>
		/// <param name="command">Command.</param>
		protected void On_ChildDoneCallback(CommandBase command)
		{
			if (On_ChildDone != null)
			{
				this.On_ChildDone (command);
			}
		}

		/// <summary>
		/// 子对象销毁时（常用于重用子对象）
		/// </summary>
		/// <param name="command">Command.</param>
		protected virtual void OnChildDestroy(CommandBase command)
		{
			command.OnDestroy ();
		}

		/// <summary>
		/// 清楚当前所有子对象
		/// </summary>
		public virtual void Clear()
		{
		}

		/// <summary>
		/// 当前对象销毁
		/// </summary>
		public override void OnDestroy ()
		{
			this.On_ChildDone = null;
			base.OnDestroy ();
		}
	}
}

File: Assets/Lua/Launch/Command/CommandContainerBase.lua
require 'Launch/Command/CommandBase'
CommandContainerBase = Class(CommandBase)

function CommandContainerBase:ctor()
	self.OnChildDoneFuncs = {}
end

--[[function CommandContainerBase:New()
	local o = CommandBase:New()
	o.OnChildDone = {}
	setmetatable(o,self)
	self.__index = self
	return o
end]]

--添加执行子命令完成的回调  格式如:func(cmd) end
function CommandBase:AddChildDoneFunc(func)
	if(not table.contains(self.OnChildDoneFuncs,func)) then
		table.insert(self.OnChildDoneFuncs,func)
	end
end
--移除子命令完成的回调
function CommandBase:RemoveChildDoneFunc(func)
	local index = -1
	for i,v in ipairs(self.OnChildDoneFuncs) do
		if(v == func) then
			index = i
		end
	end
	if(index > 0) then
		table.remove(self.OnChildDoneFuncs,index)
	end
end

function CommandContainerBase:OnChildDone(cmd)
	self:OnChildDoneInvoke(cmd)
end

function CommandContainerBase:OnChildDoneInvoke(cmd)
	if(self.OnChildDoneFuncs ~= nil) then
		for k,v in pairs(self.OnChildDoneFuncs) do
			if(v ~= nil) then
				v(cmd)
			end
		end
	end
end

function CommandContainerBase:OnExecuteDone(state)
	CommandContainerBase.superclass.OnExecuteDone(self,state)
end

--子对象销毁时（常用于重用子对象）
function CommandContainerBase:OnChildDestroy(cmd)
	cmd:OnDestroy()
end

function CommandContainerBase:OnDestroy()
	self.OnChildDoneFuncs = nil
	CommandContainerBase.superclass.OnDestroy(self)
end